### ISSUE TITLE
After deriving loss, some parameters require_grad are false

### ISSUE THREAD

**Issue by James6Chou** [facebookresearch/fairseq#5053]
#### What is your question?

I made some modifications in fairseq_task.py, specifically in the train_step() function of the FairseqTask. I computed the gradient of the loss and set retain_graph=True and create_graph=True. However, I noticed that some of the parameters in the resulting gradients had requires_grad set to True, while others had requires_grad set to False. As a result, I couldn't perform second-order differentiation. Can you help me figure out what went wrong?
I noticed that the requires_grad parameter set to False is:
Encoder layer：
encoder.layers.self_attn.out_proj.bias
encoder.layers.self_attn_layer_norm.weight
encoder.layers.self_attn_layer_norm.bias
encoder.layers.fc2.bias
encoder.layers.final_layer_norm.weight
encoder.layers.final_layer_norm.bias
Decoder layer：
decoder.layers.self_attn.out_proj.bias
decoder.layers.self_attn_layer_norm.weight
decoder.layers.self_attn_layer_norm.bias
decoder.layers.encoder_attn.out_proj.bias
decoder.layers.encoder_attn_layer_norm.weight
decoder.layers.encoder_attn_layer_norm.bias
decoder.layers.fc2.bias
decoder.layers.final_layer_norm.weight
decoder.layers.final_layer_norm.bias

#### Code

    def train_step(
        self, sample, model, criterion, optimizer, update_num, ignore_grad=False
    ):
        model.train()
        model.set_num_updates(update_num)
        with torch.autograd.profiler.record_function("forward"):
            with torch.cuda.amp.autocast(enabled=(isinstance(optimizer, AMPOptimizer))):
                loss, sample_size, logging_output,loss_up,loss_down = criterion(model, sample)#要在这里动刀了！
        if ignore_grad:
            loss *= 0
        with torch.autograd.profiler.record_function("backward"):
            grad_test = torch.autograd.grad(outputs=loss,inputs=model.parameters(),retain_graph=True,create_graph=True) #compute the gradient of the loss
            for i in range(0,len(grad_test)):
                if(grad_test[i].requires_grad==False): # requires_grad set to False
                    print(list(model.named_parameters())[i][0])

#### What's your environment?

 - fairseq Version 0.12.1
 - PyTorch Version 1.13.1
 - OS Linux
 - How you installed fairseq: source
 - Build command you used : pip install --editable /path/to/fairseq/source
 - Python version:3.8.3
 - CUDA/cuDNN version: 11.7
